src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscovery.java

```diff
@@ -46,6 +46,60 @@
         }
     }
 
+    static InstanceDiscoveryMetadataEntry GetMetadataEntry(
+            URL authority,
+            ClientApplicationBase clientApplication,
+            MsalRequest msalRequest) {
+
+        if(cache.isEmpty() && !clientApplication.tokenCache().instanceDiscoveryMetadata.isEmpty()){
+
+            clientApplication.tokenCache().lock.readLock().lock();
+            try {
+
+                log.info("Instance discovery metadata populated from token cache");
+
+                cache.putAll(clientApplication.tokenCache().instanceDiscoveryMetadata);
+            } finally {
+                clientApplication.tokenCache().lock.readLock().unlock();
+            }
+        }
+
+        InstanceDiscoveryMetadataEntry result = cache.get(authority.getAuthority());
+
+        if (result == null ||
+                result.expiresOn < TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())) {
+
+            doInstanceDiscoveryAndCache(
+                    authority,
+                    clientApplication.validateAuthority(),
+                    msalRequest,
+                    clientApplication.getServiceBundle());
+
+            clientApplication.tokenCache().saveInstanceDiscoveryMetadata(msalRequest, cache);
+        }
+
+        return cache.get(authority.getAuthority());
+    }
+
+    private static void doInstanceDiscoveryAndCache(
+            URL authorityUrl,
+            boolean validateAuthority,
+            MsalRequest msalRequest,
+            ServiceBundle serviceBundle) {
+
+        log.info("Instance Discovery information not found in cache. Sending instance discovery " +
+                "request to AAD");
+
+        InstanceDiscoveryResponse instanceDiscoveryResponse =
+                sendInstanceDiscoveryRequest(authorityUrl, msalRequest, serviceBundle);
+
+        if (validateAuthority) {
+            validate(instanceDiscoveryResponse);
```

Comment:
[Aside] The way you do `validate` is different from how .net does it so it might be a good idea to align here... Can you give me an example where the instance discovery endpoint would fail validation ? I manually tested using a "bogus authority" and the instance discovery endpoint fails with a 500 error I believe.

Reply:
We are not validating instance discovery endpoint here, only validating that the response contains "tenant_discovery_endpoint". Authority validation happens here: https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/dev/src/main/java/com/microsoft/aad/msal4j/Authority.java#L88

What validation happens in .net? Agreed that we should align here.